cpp/src/arrow/device.h

```diff
@@ -109,19 +109,50 @@ class ARROW_EXPORT Device : public std::enable_shared_from_this<Device>,
   /// should be trivially constructible from it's device-specific counterparts.
   class ARROW_EXPORT Stream {
    public:
-    virtual const void* get_raw() const { return NULLPTR; }
+    using release_fn_t = void (*)(void*);
```

Comment:
Should we consider making this `std::function` instead of a raw function pointer here? Using the former would allow for passing lambda captures at the cost of some overhead. Lambda captures would be useful for something like if someone has a smart pointer managing the lifetime of a stream.